src/vm_cpu.rs

```diff
@@ -97,6 +100,192 @@ pub struct VmCpuTrapState {
     pub htval: u64,
     pub htinst: u64,
 }
+#[derive(Default)]
+#[repr(C)]
+struct VmPmuState {
+    // Stores values of PMU CSRs
+    pmu_csrs: [u64; 32],
+    // Stores an index into PmuCounterInfo for configured counters (max index is 31)
+    configured_counters: u64,
+    started_counters: u64,
+}
+
+impl VmPmuState {
+    #[inline]
+    fn read_csr<const V: u16>() -> u64 {
+        let r: u64;
+        unsafe {
```

Comment:
why is this safe? Is reading any CSR always safe?

Reply:
Good point...it was intended to be used specifically in context of reading PMU registers (which is safe), but will change it to unsafe for the general case.

Reply:
Actually, there are other places in the code base where we read CSRs without unsafe. Anyway, I have moved it into a local function, and there's an explicit check for the CSR before the attempted read now.